### PR TITLE
osdc: additional processing of remove_snaps on the latest map

### PR DESCRIPTION
**Pull request by simon-rock**
Signed-off-by: Simon Gao <simon29rock@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ukernel** [comment]
LGTM

**Comment by batrick** [comment]
@ukernel please review again

**Review by batrick — DISMISSED** [review]
There's odd spacing in the commit message:

https://github.com/ceph/ceph/pull/36495/commits/c8014867544e537d7eb69846c6cac980804571e3

Please remove the extra spaces. Something like:

````
client: additional processing of remove_snaps on the latest map

The original processing scheme doesn't handle the removed snaps of latest osdmap.
````

**Review by ukernel — APPROVED** [review]

**Comment by simon-rock** [comment]
> ping @simon-rock

updated it. thx

**Comment by tchaikov** [comment]
will take a look early next week.

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
Passes cephfs QA: https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200924.050041

Failures known/unrelated.

@tchaikov do you still want to review this?

**Comment by batrick** [comment]
I'm investigating the some snapshot failures which may be related to this. So hold off on merging until I come back with results.

File: src/osdc/Objecter.cc [hunk 1]
```diff
@@ -1214,9 +1215,10 @@ void Objecter::handle_osd_map(MOSDMap *m)
 	update_pool_full_map(pool_full_map);
 
 	// check all outstanding requests on every epoch
```
Comment (tchaikov) [inline src/osdc/Objecter.cc:1217]: might want to update the comment accordingly, like

> collect removed snap in every past osdmaps

File: src/osdc/Objecter.cc [hunk 2]
```diff
@@ -1214,9 +1215,10 @@ void Objecter::handle_osd_map(MOSDMap *m)
...
-	for (auto& i : need_resend) {
-	  _prune_snapc(osdmap->get_new_removed_snaps(), i.second);
+	for (auto& i : osdmap->get_new_removed_snaps()) {
+	  removed_snaps[i.first].insert(i.second);
```
Comment (tchaikov) [inline src/osdc/Objecter.cc:1219]:
```suggestion
	for (auto& [pool, snaps] : osdmap->get_new_removed_snaps()) {
          removed_snaps[pool].insert(snaps);
```

**Review by batrick — CHANGES_REQUESTED** [review]
This appears to cause:

````
2020-10-08T20:36:35.097 DEBUG:tasks.ceph_test_case:wait_until_true: waiting (timeout=30.0)...
2020-10-08T20:36:38.533 INFO:tasks.ceph.mgr.y.smithi172.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-6203-g85e10386/rpm/el8/BUILD/ceph-16.0.0-6203-g85e10386/src/include/interval_set.h: In function 'void interval_set<T, C>::insert(T, T, T*, T*) [with T = snapid_t; C = mempool::osdmap::flat_map]' thread 7f016c052700 time 2020-10-08T20:36:38.526118+0000
2020-10-08T20:36:38.533 INFO:tasks.ceph.mgr.y.smithi172.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-6203-g85e10386/rpm/el8/BUILD/ceph-16.0.0-6203-g85e10386/src/include/interval_set.h: 495: FAILED ceph_assert(p->first > start+len)
2020-10-08T20:36:38.533 INFO:tasks.ceph.mgr.y.smithi172.stderr: ceph version 16.0.0-6203-g85e10386 (85e10386d1b7a7bab8cce51e6ac335c773c5c946) pacific (dev)
2020-10-08T20:36:38.533 INFO:tasks.ceph.mgr.y.smithi172.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x158) [0x7f0197e658c6]
2020-10-08T20:36:38.533 INFO:tasks.ceph.mgr.y.smithi172.stderr: 2: /usr/lib64/ceph/libceph-common.so.2(+0x275ae0) [0x7f0197e65ae0]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 3: /lib64/librados.so.2(+0xf3842) [0x7f018832a842]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 4: /lib64/librados.so.2(+0xf7353) [0x7f018832e353]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 5: /lib64/librados.so.2(+0xbc3ba) [0x7f01882f33ba]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 6: (DispatchQueue::entry()+0x126a) [0x7f019808ffaa]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 7: (DispatchQueue::DispatchThread::entry()+0x11) [0x7f01981412e1]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 8: /lib64/libpthread.so.0(+0x82de) [0x7f01962ee2de]
2020-10-08T20:36:38.534 INFO:tasks.ceph.mgr.y.smithi172.stderr: 9: clone()
````

From: /ceph/teuthology-archive/pdonnell-2020-10-08_20:03:26-multimds-wip-pdonnell-testing-20201008.013601-distro-basic-smithi/5507075/teuthology.log
or https://pulpito.ceph.com/pdonnell-2020-10-08_20:03:26-multimds-wip-pdonnell-testing-20201008.013601-distro-basic-smithi/5507075/

The same tests do not fail with master:

https://pulpito.ceph.com/pdonnell-2020-10-08_20:58:25-multimds-master-distro-basic-smithi/

(the failures in that run are fixed by #36004 which was in the branch wip-pdonnell-testing-20201008.013601 but not master)

**Comment by batrick** [comment]
Sorry I'm wrong. This also does occur on master, we had an instance here: https://pulpito.ceph.com/teuthology-2020-09-21_04:15:02-multimds-master-distro-basic-smithi/5454314/

This PR should be good to merge.

File: src/osdc/Objecter.cc (hunk 1 above)
Comment (simon-rock) [reply src/osdc/Objecter.cc]: updated it

File: src/osdc/Objecter.cc (hunk 2 above)
Comment (simon-rock) [reply src/osdc/Objecter.cc]: updated it

**Comment by simon-rock** [comment]
@jdurgin review it please, thx.

**Comment by simon-rock** [comment]
@batrick can you help me to merge it, or what else do I need to do? thx

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
@simon-rock FYI, the commit title component prefix should be "osdc: ".

**Review by idryomov — COMMENTED** [review]
I see that Zheng approved this and several people left nits not related to the logic of the patch, which means that the patch itself made sense to them.  Since Patrick tagged me, I stared at this for a bit and I don't understand what the issue is.  Can someone explain how "the removed snaps of latest osdmap" aren't handled?

IIUC `_prune_snapc()` is supposed to be called for each outstanding op, for each osdmap.  When processing a given osdmap, there are two places to look for outstanding ops: `need_resend` and OSD sessions (initially all outstanding ops are linked into their OSD sessions, but, as osdmaps get processed, some ops get unlinked from their OSD session and added to `need_resend`).  `_prune_snapc()` call in `handle_osd_map()` takes care of the ops in `need_resend` and `_prune_snapc()` call in `_scan_requests()` takes care of the remaining OSD session ops.  There isn't anything special about the last osdmap.  What am I missing?

**Pull request closed without merging**